src/algokit/core/compile.py

```diff
@@ -0,0 +1,80 @@
+from collections.abc import Iterator
+
+from algokit.core.proc import run
+from algokit.core.utils import extract_version_triple, find_valid_pipx_command
+
+
+def find_valid_puyapy_command(version: str | None) -> list[str]:
+    return _find_puya_command_at_version(version) if version is not None else _find_puya_command()
+
+
+def _find_puya_command_at_version(version: str) -> list[str]:
+    """
+    Find puya command with a specific version.
+    If the puya version isn't installed, install it with pipx run.
+    """
+    for puyapy_command in _get_candidates_puyapy_commands():
+        try:
+            puyapy_version_result = run([*puyapy_command, "--version"])
+        except OSError:
+            pass  # in case of path/permission issues, go to next candidate
+        else:
+            if puyapy_version_result.exit_code == 0 and (
+                extract_version_triple(version) == extract_version_triple(puyapy_version_result.output)
+            ):
+                return puyapy_command
+
+    pipx_command = find_valid_pipx_command(
+        "Unable to find pipx install so that `PuyaPy` compiler can be installed; "
+        "please install pipx via https://pypa.github.io/pipx/ "
+        "and then try `algokit compile py ...` again."
+    )
+
+    return [
+        *pipx_command,
```

Comment:
Does pipx check python version constraints when running or installing a dependency?
It would be good to confirm we gracefully handle when the user is running Python < 3.12.

Poetry I think is a different story, as it should use the python in their venv, which I'm assuming should be 3.12 if they've installed Puya.

Reply:
`pipx install puya` throws an exception if the python version on the user's machine is < 3.12. It isn't easy to handle it nicely because we don't know which python version is used by pipx. I suggest that we throw the pipx exception (the current behaviour).